### PR TITLE
IGVF-586-taxa-calculated-property-new

### PR DESCRIPTION
**Pull request by corismall**

### CODE REVIEW

File: src/igvfd/tests/test_audit_biosample.py [hunk 1]
```diff
@@ -21,3 +21,34 @@ def test_audit_biosample_nih_institutional_certification(
...
+    res = testapp.get(tissue['@id'] + '@@index-data')
+    assert not any(
+        error['category'] == 'inconsistent/mixed donor taxa'
+        for error in res.json['audit'].get('ERROR', [])
+    )
+    testapp.patch_json(tissue['@id'],
+                       {'donors': [human_donor['@id']]}
+                       )
+    res = testapp.get(tissue['@id'] + '@@index-data')
+    assert not any(
```
Comment (gabdank) [inline src/igvfd/tests/test_audit_biosample.py:51]: I think you can write it as: `assert all(error != 'inconsistent)`

File: src/igvfd/tests/test_audit_biosample.py [hunk 2]
```diff
@@ -21,3 +21,34 @@ def test_audit_biosample_nih_institutional_certification(
...
+    assert any(
+        error['category'] == 'inconsistent/mixed donor taxa'
+        for error in res.json['audit'].get('ERROR', [])
+    )
+    testapp.patch_json(tissue['@id'],
+                       {'donors': [human_donor['@id'],
+                                   human_male_donor['@id']]}
+                       )
+    res = testapp.get(tissue['@id'] + '@@index-data')
+    assert not any(
```
Comment (gabdank) [inline src/igvfd/tests/test_audit_biosample.py:43]: I think you can write it as: `assert all(error != 'inconsistent)`

File: src/igvfd/audit/biosample.py [hunk 3]
```diff
@@ -18,3 +18,27 @@ def audit_biosample_nih_institutional_certification(value, system):
...
+                if taxa not in taxa_dict:
+                    taxa_dict[taxa] = []
+
+                taxa_dict[taxa].append(d)
+
+        if len(taxa_dict) > 1:
+            detail = ''
+            for k, v in taxa_dict.items():
+                detail += f'Biosample {audit_link(sample_id, sample_id)} has donors {audit_link(v, v)} that are {k}. '
+            yield AuditFailure('inconsistent/mixed donor taxa', detail, level='ERROR')
```
Comment (gabdank) [inline src/igvfd/audit/biosample.py:44]: please rename the category to be just "inconsistent donor taxa"

File: src/igvfd/upgrade/sample.py [hunk 4]
```diff
@@ -166,3 +166,20 @@ def in_vitro_system_7_8(value, system):
...
+@upgrade_step('primary_cell', '11', '12')
+@upgrade_step('tissue', '11', '12')
+@upgrade_step('whole_organism', '10', '11')
+@upgrade_step('in_vitro_system', '8', '9')
+def biosample_7_8(value, system):
+    # https://igvf.atlassian.net/browse/IGVF-586
+    if 'taxa' in value:
+        if 'notes' in value:
+            notes_value = value['notes']
+            if len(notes_value) > 0:
```
Comment (gabdank) [inline src/igvfd/upgrade/sample.py:180]: can be just: "if notes_value:"

**Review by gabdank — CHANGES_REQUESTED** [review]
minor suggestions

**Review by gabdank — APPROVED** [review]

File: src/igvfd/schemas/biosample.json [hunk 5]
```diff
@@ -56,39 +55,6 @@
                 "lower_bound_age",
                 "age_units"
             ]
-        },
-        "age_units": {
-            "comment": "Human biosamples cannot have age units of minutes or hours. Yeast biosamples cannot have age units of years or months.",
-            "required": [
-                "lower_bound_age",
-                "upper_bound_age"
```
Comment (ian-whaling) [inline src/igvfd/schemas/biosample.json:64]: Why was this removed? I think "lower_bound_age" and "upper_bound_age" should still be required for submitting "age_units."

File: src/igvfd/tests/test_upgrade_in_vitro_system.py [hunk 6]
```diff
@@ -38,11 +38,18 @@ def test_in_vitro_system_upgrade_5_6(upgrader, in_vitro_system_v5):
 def test_in_vitro_system_upgrade_6_7(upgrader, in_vitro_system_v6):
     value = upgrader.upgrade('in_vitro_system', in_vitro_system_v6, current_version='6', target_version='7')
     assert value['schema_version'] == '7'
-    assert value['taxa'] != 'Saccharomyces'
+    assert value['taxa'] == 'Saccharomyces'
```
Comment (ian-whaling) [inline src/igvfd/tests/test_upgrade_in_vitro_system.py:41]: Should this be removed like the rest of the upgrades?

File: src/igvfd/schemas/biosample.json (hunk 5 above)
Comment (corismall) [reply src/igvfd/schemas/biosample.json:64]: you're right, i re-added them

File: src/igvfd/tests/test_upgrade_in_vitro_system.py (hunk 6 above)
Comment (corismall) [reply src/igvfd/tests/test_upgrade_in_vitro_system.py]: also removed taxa here, thanks for pointing that out

File: src/igvfd/tests/test_types_biosample.py [hunk 7]
```diff
@@ -155,3 +155,21 @@ def test_summary(testapp, tissue, primary_cell, whole_organism, in_vitro_cell_li
...
+    res = testapp.patch_json(
+        tissue['@id'],
+        {'donors': [rodent_donor['@id']]})
+    res = testapp.get(tissue['@id'])
+    assert res.json.get('taxa') == 'Mus musculus'
+    res = testapp.patch_json(
+        tissue['@id'],
+        {'donors': [human_donor['@id'], rodent_donor['@id']]})
+    res = testapp.get(tissue['@id'])
+    assert res.json.get('taxa') == None
```
Comment (gabdank) [inline src/igvfd/tests/test_types_biosample.py:175]: I think  this should be `res.json.get('taxa', None) == None`

**Review by gabdank — CHANGES_REQUESTED** [review]
one minor comment

File: src/igvfd/schemas/changelogs/whole_organism.md [hunk 8]
```diff
@@ -1,37 +1,40 @@
-## Changelog for *`whole_organism.json`*
+# Changelog for *`whole_organism.json`*
```
Comment (gabdank) [inline src/igvfd/schemas/changelogs/whole_organism.md:1]: why did you modify all the "#" ? I think you should stay consistent with all the other changelogs

**Review by gabdank — CHANGES_REQUESTED** [review]
update the # in whole organism changelog

File: src/igvfd/schemas/changelogs/whole_organism.md (hunk 8 above)
Comment (corismall) [reply src/igvfd/schemas/changelogs/whole_organism.md:1]: i was having issues with linting following rebase/resolution of conflicts, autopep8 wasn't passing so I had to run it on several files from the command line before it passed. This was one of the files.

File: src/igvfd/schemas/changelogs/whole_organism.md [hunk 9]
```diff
@@ -51,7 +54,7 @@
 
 ### Schema version 3
 
-* Restrict `treatments`, `donors`, `dbxrefs`, `aliases`, `collections`, and `alternate_accessions` to be a non-empty array with at least one item.
+* Restrict `treatments`, `donors`, `dbxrefs`, `aliases`, `collections`, and `alternate_accessions` to be a non - empty array with at least one item.
```
Comment (ian-whaling) [inline src/igvfd/schemas/changelogs/whole_organism.md:57]: I think this should remain the same

**Review by gabdank — APPROVED** [review]

**Review by gabdank — APPROVED** [review]
